docs/src/communication.md

```diff
@@ -0,0 +1,202 @@
+# [Communication between Julia and Javascript](@id Communication)
+
+After creating a Window and loading HTML and JS, you may want to interact with
+julia code (e.g. by clicking a button in HTML, or displaying a plot from julia).
+
+This section covers this two-way communication.
+
+## Julia to Javascript
+```@setup Blink-win
+using Blink
+win = Window(Dict(:show=>false), async=false)
+```
+
+The easiest way to communicate to javascript from julia is with the [`@js`](@ref) and
+[`@js_`](@ref) macros. These macros allow you to execute arbitrary javascript code in a
+given Window.
+
+```@repl Blink-win
+@js win x = 5;
+@js win x
+```
+
+The `@js_` macro executes its code asynchronously, but doesn't return its
+result:
+```@repl Blink-win
+@time @js win begin   # Blocks until finished; `i` is returned
+  for i in 0:1000000 end  # waste time
+  i  # return i
+end
+
+@time @js_ win begin   # Returns immediately, but `i` is not returned.
+  for i in 0:1000000 end  # waste time
+  i  # This is ignored
+end
+```
+
+If your javascript expression is complex, or you want to copy-paste existing
+javascript, it can be easier to represent it as a pure javascript string.
+For that, you can call the [`js`](@ref) function with a [`JSString`](@ref Blink.JSString):
+```@repl Blink-win
+body!(win, """<div id="box" style="color:red;"></div>""", async=false);
+div_id = "box";
+js(win, Blink.JSString("""document.getElementById("$div_id").style.color"""))
+```
+
+Note that the code passed to these macros runs in its own scope, so any
+javascript variables you create with `var` (or the `@var` equivalent for `@js`)
+will be inaccessible after returning:
+```@repl Blink-win
+@js win (@var x_var = 5; x_var)  # x_var is only accessible within this scope.
+@js win x_var
+```
+
+## Javascript to Julia
+Communication from javascript to julia currently works via a message passing
+interface.
+
+To invoke julia code from javascript, you specify a julia callback via `handle`:
+```julia-repl
+julia> handle(w, "press") do args
+         @show args
+       end
+```
+This callback can then be triggered from javscript via `Blink.msg()`:
+```@setup handler
+using Blink
+w = Window(Dict(:show=>false), async=false)
+handle(w, "press") do args
+  @show args
+end
+```
+```@repl handler
+@js w Blink.msg("press", "Hello from JS");
+```
+Note that `Blink.msg` takes _exactly_ 1 argument.  To pass more or fewer,
+consider passing an array:
+```@repl handler
+@js w Blink.msg("press", [1,2,3]);
+```
+(TODO: is this on purpose? Can we fix this to allow variadic args?)
```

Comment:
_Is_ this on purpose? Maybe I should start a separate issue about it, but it seems weird that this is the API. Could the javascript function be changed to be variadic? (I actually don't know if you can do that in javascript..)